### PR TITLE
Fixed #17688 -- Added actions to m2m_changed for deleted objects.

### PR DESCRIPTION
**Pull request by jorgecarleitao**
This is an update to the PR #5505. I have implemented the changes in the comments of that PR.

### CODE REVIEW

**Comment by auvipy** [comment]
@jorgecarleitao still contains merge conflicts. probably you should rebase again :)

**Comment by jorgecarleitao** [comment]
Thanks @auvipy. It was a minor thing on the docs. I've just rebased it.

**Comment by tomchristie** [comment]
@auvipy I understand it's well intentioned, but these sorts of comments on issues ("needs rebasing", "will this be in the next version" etc) aren't terribly constructive - on balance it's probably better to leave them out. Some folks won't mind, or might even find them useful, but I'd expect others may not appreciate them. Not a big deal, but figure worth mentioning it, because I see these fairly frequently from you.

**Comment by phildini** [comment]
@tomchristie: do you feel the comment changes need to be made before accepting?

**Comment by tomchristie** [comment]
@phildini Assuming you mean this?... https://github.com/django/django/pull/5505#discussion-diff-51617233R300

No particular opinion, it could be a little clearer, but I think I understand the intent given the existing comment wording.

**Comment by jorgecarleitao** [comment]
If anyone has a better rewording, please post here and I update the commit.

File: tests/m2m_signals/tests.py [hunk 1]
```diff
@@ -461,3 +461,55 @@ def test_m2m_relations_with_self_add_idols(self):
...
+            'objects': [self.bmw, self.toyota, self.vw],
+        }]
+        self._initialize_signal_car(add_default_parts_before_set_signal=False)
+        self.airbag.delete()
+
+        self.assertEqual(self.m2m_changed_messages, expected_messages)
+
+    def test_m2m_relations_delete_reverse_related(self):
+        """
+        Tests that deleting an element belonging to a reversed m2m triggers
```
Comment (timgraham) [inline tests/m2m_signals/tests.py]: as above, chop "Tests that"

File: tests/m2m_signals/tests.py [hunk 2]
```diff
@@ -461,3 +461,55 @@ def test_m2m_relations_with_self_add_idols(self):
...
+        }, {
+            'instance': self.airbag,
+            'action': 'post_delete',
+            'reverse': True,
+            'model': Car,
+            'objects': [self.bmw, self.toyota, self.vw],
+        }]
+        self._initialize_signal_car(add_default_parts_before_set_signal=False)
+        self.airbag.delete()
+
```
Comment (timgraham) [inline tests/m2m_signals/tests.py]: chop blank line

File: tests/m2m_signals/tests.py [hunk 3]
```diff
@@ -461,3 +461,55 @@ def test_m2m_relations_with_self_add_idols(self):
...
             }
         ])
+
+    def test_m2m_relations_delete_related(self):
+        """
+        Deleting an element belonging to a m2m triggers the signals
+        `pre_delete` and `post_delete`.
+        """
+        self.airbag.cars_optional.add(self.vw, self.bmw, self.toyota)
+
```
Comment (timgraham) [inline tests/m2m_signals/tests.py]: chop blank line

File: tests/m2m_signals/tests.py [hunk 4]
```diff
@@ -461,3 +461,55 @@ def test_m2m_relations_with_self_add_idols(self):
...
+        self.assertEqual(self.m2m_changed_messages, expected_messages)
+
+    def test_m2m_relations_delete_reverse_related(self):
+        """
+        Tests that deleting an element belonging to a reversed m2m triggers
+        the signals `pre_delete` and `post_delete`.
+        """
+        self.vw.optional_parts.add(self.airbag, self.sunroof)
+
```
Comment (timgraham) [inline tests/m2m_signals/tests.py]: chop blank line

File: tests/m2m_signals/tests.py [hunk 5]
```diff
@@ -461,3 +461,55 @@ def test_m2m_relations_with_self_add_idols(self):
...
+            'model': Part,
+            'objects': [self.airbag, self.sunroof],
+        }, {
+            'instance': self.vw,
+            'action': 'post_delete',
+            'reverse': True,
+            'model': Part,
+            'objects': [self.airbag, self.sunroof],
+        }]
+
```
Comment (timgraham) [inline tests/m2m_signals/tests.py]: chop blank line (and below too)

File: docs/releases/1.10.txt [hunk 6]
```diff
@@ -452,6 +452,9 @@ Models
   <django.db.models.Options.default_manager_name>` options allow controlling
   the :attr:`~django.db.models.Model._base_manager` and
   :attr:`~django.db.models.Model._default_manager`, respectively.
+* When an object in a :class:`~django.db.models.ManyToManyField` is deleted,
```
Comment (timgraham) [inline docs/releases/1.10.txt]: move to 1.11

File: docs/ref/signals.txt [hunk 7]
```diff
@@ -217,6 +217,11 @@ instance. Strictly speaking, this is not a model signal since it is sent by the
 :data:`pre_save`/:data:`post_save` and :data:`pre_delete`/:data:`post_delete`
 when it comes to tracking changes to models, it is included here.
 
+.. versionadded:: 1.10
```
Comment (timgraham) [inline docs/ref/signals.txt]: 1.11 and move to the bottom of the "action" section where the other additions are made

File: django/db/models/deletion.py [hunk 8]
```diff
@@ -279,6 +329,8 @@ def delete(self):
                         sender=model, instance=obj, using=self.using
                     )
```
Comment (timgraham) [inline django/db/models/deletion.py]: chop blank line here and in the similar code below (the parenthesis on its own line provides whitespace)

File: django/db/models/deletion.py [hunk 9]
```diff
@@ -271,6 +271,56 @@ def delete(self):
         # number of objects deleted for each model label
         deleted_counter = Counter()
 
+        # caches query results of signals to avoid one hit per ['pre_', 'post_']
+        cached_pk_sets = dict()
+
+        def send_m2m_changed(model, obj, action):
+            """
+            Sends `m2m_changed` to all m2m relationships of this `model` that
```
Comment (timgraham) [inline django/db/models/deletion.py]: use PEP257 verb style: "Send"

File: django/db/models/deletion.py [hunk 10]
```diff
@@ -271,6 +271,56 @@ def delete(self):
...
+            contain at least one row in the table.
+            """
+            m2m_relations = [
+                f for f in model._meta.get_fields(include_hidden=True)
```
Comment (timgraham) [inline django/db/models/deletion.py]: could you add a test that fails if `include_hidden=True` is removed?

File: django/db/models/deletion.py [hunk 11]
```diff
@@ -271,6 +271,56 @@ def delete(self):
...
+                if f.many_to_many
+            ]
+
+            from django.db.models import ManyToManyField, ManyToManyRel
+            # for each relation, get all ids of the other side of
```
Comment (timgraham) [inline django/db/models/deletion.py]: Use capitalization and periods

File: django/db/models/deletion.py [hunk 12]
```diff
@@ -271,6 +271,56 @@ def delete(self):
...
+            """
+            Sends `m2m_changed` to all m2m relationships of this `model` that
+            contain at least one row in the table.
+            """
+            m2m_relations = [
+                f for f in model._meta.get_fields(include_hidden=True)
+                if f.many_to_many
+            ]
+
+            from django.db.models import ManyToManyField, ManyToManyRel
```
Comment (timgraham) [inline django/db/models/deletion.py]: put import at the top of the method

File: django/db/models/deletion.py [hunk 13]
```diff
@@ -271,6 +271,56 @@ def delete(self):
...
+            # for each relation, get all ids of the other side of
+            # the relation and send the signal with them
+            for m2m_relation in m2m_relations:
+
```
Comment (timgraham) [inline django/db/models/deletion.py]: chop blank line

File: django/db/models/deletion.py [hunk 14]
```diff
@@ -271,6 +271,56 @@ def delete(self):
...
+                # fixme: find a simpler idiom to select the manager and sender.
```
Comment (timgraham) [inline django/db/models/deletion.py]: fixme -> TODO

File: django/db/models/deletion.py [hunk 15]
```diff
@@ -271,6 +271,56 @@ def delete(self):
...
+                # ATM is with isinstance and getattr(obj, attr_name):
```
Comment (timgraham) [inline django/db/models/deletion.py]: ATM is -> than with

File: django/db/models/deletion.py [hunk 16]
```diff
@@ -271,6 +271,56 @@ def delete(self):
...
+                if isinstance(m2m_relation, ManyToManyField):
+                    attr_name = m2m_relation.name
+                    sender = m2m_relation.remote_field.through
+                elif isinstance(m2m_relation, ManyToManyRel):
+                    attr_name = m2m_relation.get_accessor_name()
+                    sender = m2m_relation.through
+                else:
+                    # many_to_many fields from 3rd parties are ignored.
```
Comment (timgraham) [inline django/db/models/deletion.py]: Third-party many_to_many fields are ignored. (add an explanation of why?)

File: django/db/models/deletion.py [hunk 17]
```diff
@@ -271,6 +271,56 @@ def delete(self):
...
+                    continue
+
+                # don't send if no one is listening to avoid a db hit
+                if not signals.m2m_changed.has_listeners(sender):
+                    continue
+
+                key = (obj, attr_name)
+                if key not in cached_pk_sets:
+                    manager = getattr(obj, attr_name)
+
```
Comment (timgraham) [inline django/db/models/deletion.py]: chop blank line

File: django/db/models/deletion.py (hunk 10 above)
Comment (jorgecarleitao) [inline django/db/models/deletion.py]: Can you elaborate the rational? `include_hidden` is not a parameter.

Comment (timgraham) [inline django/db/models/deletion.py]: What I mean is that no tests fail if I change this line to `f for f in model._meta.get_fields()`

**Comment by amw** [comment]
I think it's an unnecessary complication that the events sent in this PR are "pre_delete" and "post_delete". This should be just regular "pre_remove" and "post_remove" as if the item was simply removed from the relation.

Pizza has a set of toppings. Before a topping is deleted it is removed from all the pizzas. Probably in 99% cases the pizza will not care whether the topping has been entirely deleted from the store or just removed from this particular pizza.

Having this new, special event names will complicate code in every m2m_changed hook. We just want to do something about the pizza now that this topping is no longer on it: maybe update its price.

In either case the order of events should be:
- send "pre" hook
- delete rows from m2m table
- send "post" hook
- delete the related model

**Comment by timgraham** [comment]
Thanks for the input. I don't know if someone might have a need to distinguish the delete case vs. the remove case. In any case, it would be better to make that comment on the ticket as design decisions are more often discussed there and it's easier to review a ticket than to review a PR and all its comments. Anyway, I'll close this PR for now since it looks like you plan to work on a new one.

**Comment by amw** [comment]
I am definitively going to work on top of @jorgecarleitao commit and see where it takes me. I'll re-post my comments on Trac.

**Comment by mpasternak** [comment]
@amw @timgraham @jorgecarleitao any updates on this? Looks like I'll need this. On the other hand, the issue is open and silent since 2 yrs or so. Have you implemented it differently in your apps?

**Comment by amw** [comment]
@mpasternak I've simply overridden the parent model's `delete` method to perform the work related to clearing the children. Like this:

```python
   def delete(self, **options):
       """
       Quick and dirty fix for until Django's #17688 bug is fixed.
       """
       clean_up_toppings_that_belong_to_this_pizza(self)
       return super().delete(**options)
```

**Comment by mpasternak** [comment]
@amw I have a record -> authors relationship, the record description generation is slow so I got away with caching it... now it gets rebuilt every time you add a new author... with authors up to 300-500. Same for deletion. I thought I could get away with some smart signals but it looks like that I'll need to patch admin or something. Thanks.

**Pull request closed without merging**